### ISSUE TITLE
Feature request: add support for more customisations of Draftail

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#5580]
Previously at #4315 / #4436. Related: #4445, #4223.

At the moment, rich text extensions can only customise the following parts of Draftail:

- Blocks
- Inline styles
- Entities (e.g. links, images, embeds)
- Horizontal rules being enabled or not

Here are things that are not customisable and would be very valuable to make customisable:

```js
// Displayed when the editor is empty. Hidden if the user changes styling.
placeholder: "Write here",
// Show undo control in the toolbar.
showUndoControl: true
// Show redo control in the toolbar.
showRedoControl: true,
// List of active decorators.
decorators: [],
// Additional React components to render in the toolbar.
controls: [],
// List of plugins of the draft-js-plugins architecture.
plugins: [],
// Max level of nesting for list items. 0 = no nesting. Maximum = 10.
maxListNesting: 4,
```

And things that I think would be valuable for _some_ extensions, but probably not that valuable to make customisable just yet, except to allow for experimentation:

```
// Called when the editor receives focus.
onFocus: null,
// Called when the editor loses focus.
onBlur: null,
// Optionally override the default Draftail toolbar, removing or replacing it.
topToolbar: Toolbar,
// Optionally add a custom toolbar underneath the editor, e.g. for metrics.
bottomToolbar: null,
```

---

Additionally, for those customisations that would be done with Wagtail’s rich text features, I think it’ll be important that a single rich text feature can map to multiple Draftail extensions – for example, a `code-block` rich text feature could configure both a block type for code (as is supported already) and a decorator that would do syntax highlighting. I think this might be possible already in the rich text features API, but the Draftail integration of this might not support it.

## Implementation

I think the options described above fall in different categories:

- Things that map well to "rich text features" or "rich text formatting" because they determine the formatting available in the editor in a widget-agnostic way. For example an "emoji" feature that maps to a plugin and a control, and we would likely want to enable either site-wide, or per-field
- Things that would ideally map to field or widget attributes, because they aren’t widget-agnostic. For example providing a custom placeholder string for some fields, or hiding the undo/redo controls site-wide.
- It might also make sense to have customisation APIs that are client-side only – for more advanced customisations that we might not want to have APIs for, but where it would still be valuable to reuse existing code.

As a way to assess API designs, here are types of editor extensions that I would like those new APIs to support:

- Indent/dedent buttons. This is the only operation currently doable with the keyboard only, thus unusable on touch screens with virtual keyboards that generally don’t have a "tab" key.
- Paste from Word button – content import features 
- Clear formatting button – Resets the editor to plain-text
- Character replacements – smart quotes, convert `-->` to `→` like Google Docs
- Slash-commands – interact with the editor, and Wagtail, by typing CLI-like commands with autocomplete
- Emojis – emoji picker, copy-paste support
- Content linting – validation rules, highlighted words
- Content metrics – live-updating measures of the content’s readability, SEO-keywords-ratio, sentiment analysis.
- Single-line input – the editor constrained to a single block, with no block-level formats in the toolbar. Pasting causes the content to be added in a single line.

**Comment by thibaudcolas** [comment]
For anyone looking into this, the temporary workaround for this is to monkey patch / override Draftail’s `initEditor` to implement your own API to register further customisations of Draftail. 

Here is a good example from @v9Chris: https://github.com/vixdigital/wagtail-readinglevel/blob/master/wagtailreadinglevel/static/api-monkeypatch.js

```js
// Implement the new APIs.

const DECORATORS = [];
const CONTROLS = [];

const registerDecorator = (decorator) => {
    DECORATORS.push(decorator);
    return DECORATORS;
};

const registerControl = (control) => {
    CONTROLS.push(control);
    return CONTROLS;
};

// Override the existing initEditor to hook the new APIs into it.
// This works in Wagtail 2.0 but will definitely break in a future release.
const initEditor = window.draftail.initEditor;

const initEditorOverride = (selector, options, currentScript) => {
    const overrides = {
        decorators: DECORATORS,
        controls: CONTROLS,
    };

    const newOptions = Object.assign({}, options, overrides);

    return initEditor(selector, newOptions, currentScript);
};

window.draftail.registerControl = registerControl;
window.draftail.registerDecorator = registerDecorator;
window.draftail.initEditor = initEditorOverride;
```

This is obviously not a documented / official API, so if you’re in need of this please comment on this issue to make sure we’re aware of your use case. **At some stage this implementation-specific workaround will likely go away, and we likely won’t provide any migration path or breaking change notice for this**.

**Comment by kmohrf** [comment]
I’m interested in this and have used the same approach as above to inject some functionality into the draftail instances. In my case I wanted to replace ASCII style quote-characters with [guillemets](https://en.wikipedia.org/wiki/Guillemet), which can be solved by a relatively simple draft-js plugin. Unfortunately there is no way to add draft-js plugins through the Wagtail API.

**Comment by thibaudcolas** [comment]
Thanks @kmohrf, that’s good to know. Do you have any code to share for reference? There aren’t that many people publicly talking about customizations like this, so any real-world implementation is appreciated.

For this type of character replacement – it’s something I would like to support out of the box eventually (https://github.com/springload/draftail/issues/113). As you say though, it’s also desirable to be able to add with draft-js-plugins.

**Comment by kmohrf** [comment]
I’ve published the code [as a gist](https://gist.github.com/kmohrf/b0e070b1f3e2e32f7577ed1bf3bfa863) for reference :)

**Comment by adsee42** [comment]
Hi, any update on this?
I would love to have an "add emoji" button!

**Comment by tombola** [comment]
It looks like (as predicted), the monkeypatch suggested by @thibaudcolas [above](#issuecomment-574693649) may no longer be working with current versions of wagtail, as per this issue from the wagtail-readinglevel package (that the above example was taken from).

https://github.com/wagtail/wagtail/issues/5580#issuecomment-574693649

**Comment by tombola** [comment]
This example from @alxbridge updates the approach for more recent wagtail versions
https://github.com/vixdigital/wagtail-readinglevel/pull/32

**Comment by thibaudcolas** [comment]
Clever! This approach seems to rely on defining those extensions as entities, whose JS files will then be loaded as configured.  This seems to then register an entity server-side, and a `control` client-side using the API monkeypatch. I was surprised this doesn’t result in two buttons in the editor toolbar, but I guess this might be working since [Draftail v1.4.0](https://github.com/springload/draftail/blob/main/CHANGELOG.md#v140) ("Fix empty buttons appearing when providing custom formats without a defined label or icon"). I believe we’re only using that release in Wagtail starting with Wagtail v2.13 – in older versions, this would result in two buttons in the toolbar, one of which would either break the editor when clicked, or do nothing.

So – seems pretty good to me, with the caveat that this is again relying on implementation details of existing APIs rather than something we intend to support.

**Comment by TWOC-WHT** [comment]
[Sharing more info of our use case](https://wagtailcms.slack.com/archives/C80KTJKKM/p1644313904861229) of Draftail customisation.

One of these is using a slightly modified version of the word count:

- Our contents are dual language (Chinese and English), and in Chinese, one characters equals to one word (so we need to extend the default counting via space between words)
- In addition to that we also have Total Word counts! (Excluding some fields that we don't need to count) We placed it near the save button below, neat placement I would say. This was requested by our editors, since we need the total word counts for freelance writers.

![image](https://user-images.githubusercontent.com/47908628/158938469-bae0fd19-9a9c-4374-acae-add4ce1531c1.png)

---

Oh, and also, we managed to do a quote sign replacement! (from `''` and `""`,  to `‘’` and `“”`).  
_Not included in this example to keep the sharing simple for now_
